connect/runtime/src/main/java/org/apache/kafka/connect/util/ConnectUtils.java

```diff
@@ -243,4 +249,29 @@ public static Map<String, String> patchConfig(
         });
         return result;
     }
+
+    /**
+     * Generate a deterministic hash of the supplied config. For configurations
+     * with identical key-value pairs, this hash will always be the same.
+     * @param config the config to hash; may be null
+     * @return a hash of the config
+     */
+    public static int configHash(Map<String, String> config) {
+        if (config == null)
+            return 0;
+
+        Map<String, String> toHash = new TreeMap<>(config);
+
+        byte[] serialized;
+        try {
+            serialized = OBJECT_MAPPER.writeValueAsBytes(toHash);
```

Comment:
AbstractMap (superclass of TreeMap) has a hashCode implementation which depends on the keys and values in the map. Did you consider using that method, and reject it? It looks like it could have fewer memory allocations and copying.

Reply:
I didn't want to rely on hash code implementations being identical across JVM restarts or even different JVMs. If there's an official spec somewhere that applies to all of our supported Java versions that guarantees that `AbstractMap::hashCode` or some other hash implementation will be the same everywhere, then I agree it'd be better to use that.

Reply:
They do document the hashCode implementation very precisely: https://docs.oracle.com/en%2Fjava%2Fjavase%2F21%2Fdocs%2Fapi%2F%2F/java.base/java/util/AbstractMap.html#hashCode() 

> Returns the hash code value for this map. The hash code of a map is defined to be the sum of the hash codes of each entry in the map's entrySet() view. This ensures that m1.equals(m2) implies that m1.hashCode()==m2.hashCode() for any two maps m1 and m2, as required by the general contract of [Object.hashCode()](https://docs.oracle.com/en%2Fjava%2Fjavase%2F21%2Fdocs%2Fapi%2F%2F/java.base/java/lang/Object.html#hashCode()).

https://docs.oracle.com/en%2Fjava%2Fjavase%2F21%2Fdocs%2Fapi%2F%2F/java.base/java/util/Map.Entry.html#hashCode()

> Returns the hash code value for this map entry. The hash code of a map entry e is defined to be:
> ```
>     (e.getKey()==null   ? 0 : e.getKey().hashCode()) ^
>     (e.getValue()==null ? 0 : e.getValue().hashCode())
> ```

https://docs.oracle.com/en%2Fjava%2Fjavase%2F21%2Fdocs%2Fapi%2F%2F/java.base/java/lang/String.html#hashCode()
> Returns a hash code for this string. The hash code for a String object is computed as
> ```
> s[0]*31^(n-1) + s[1]*31^(n-2) + ... + s[n-1]
> ```

I sincerely doubt that a future JDK version would change the hash algorithm of the String, but I don't know that for sure.

I am more confident about the AbstractMap implementation staying the same though: because it's an interface that can be extended, the equals and hashCode implementations are expected to hold across distinct implementations. If the AbstractMap implementation changed, I think that would break some existing implementations.